### ISSUE TITLE
Regression in functions selecting a fused type using a pointer argument

### ISSUE THREAD

**Issue by anntzer** [cython/cython#3881]
**Describe the bug**
Cythonization of the (simplified) example below fails with Cython 0.29.21 and 3.0a6, but used to work with e.g. Cython 0.29.0.

**To Reproduce**
Code to reproduce the behaviour:
```cython
#cython: language_level=3
cimport cython
from libc.stdlib cimport malloc


cdef cython.numeric[::1] mview(size_t n, cython.numeric* null) except *:
    return <cython.numeric[:n:1]>(malloc(n * sizeof(null[0])))


def main(size_t n):
    cdef double d
    return mview(n, &d)
```
This fails with
```
Error compiling Cython file:
------------------------------------------------------------
...
cimport cython
from libc.stdlib cimport malloc


cdef cython.numeric[::1] mview(size_t n, cython.numeric* null) except *:
    return <cython.numeric[:n:1]>(malloc(n * sizeof(null[0])))
          ^
------------------------------------------------------------

foo.pyx:7:11: Type is not specialized
```

**Expected behavior**
The goal is to have a function (`mview`) that returns a newly allocated memoryview with a type matching the type of the pointer passed as second argument (the second argument is basically used as the equivalent of a template parameter in C++).
This works with Cython 0.29.0, but not with the latest release or prerelease.

**Environment (please complete the following information):**
 - OS: Linux
 - Python version: 3.8.4
 - Cython version: 0.29.21/3.0a6

**Additional context**

**Comment by da-woods** [comment]
It looks to be caused by 91eb518096d4f63d2db5a6e5140a70aecc0bc70f. It does look like it should work.

**Comment by scoder** [comment]
Yes, the fix for #3642 might have been the culprit here.

**Comment by anntzer** [comment]
thanks!
